core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardsStoresAction.java

```diff
@@ -0,0 +1,255 @@
+                            shardStoreStatuses.add(response);
+                        }
+                    }
+                    indexShardsBuilder.put(res.shardId.id(), shardStoreStatuses);
+                    shardsResponses.put(res.shardId.getIndex(), indexShardsBuilder);
+                    for (FailedNodeException failure : res.failures) {
+                        failures.add(new IndicesShardsStoresResponse.Failure(failure.nodeId(), res.shardId.getIndex(), res.shardId.id(), failure.getCause()));
+                    }
+                }
+                ImmutableOpenMap.Builder<String, ImmutableOpenIntMap<java.util.List<IndicesShardsStoresResponse.ShardStoreStatus>>> indicesShardStoreStatuses = ImmutableOpenMap.builder();
+                for (Map.Entry<String, IntObjectHashMap<java.util.List<NodeGatewayStartedShards>>> entry : shardsResponses.entrySet()) {
+                    String index = entry.getKey();
+                    ImmutableOpenIntMap.Builder<java.util.List<IndicesShardsStoresResponse.ShardStoreStatus>> shardStoreResponses = ImmutableOpenIntMap.builder();
+                    for (IntObjectCursor<java.util.List<NodeGatewayStartedShards>> shardResponse : entry.getValue()) {
+                        java.util.List<IndicesShardsStoresResponse.ShardStoreStatus> shardStoreStatusList = new ArrayList<>(shardResponse.value.size());
+                        for (int i = 0; i < shardResponse.value.size(); i++) {
+                            NodeGatewayStartedShards res = shardResponse.value.get(i);
+                            final IndicesShardsStoresResponse.ShardStoreStatus.Allocation allocation;
+                            if (res.storeException() == null) {
+                                allocation = isPrimary(index, shardResponse.key, res.getNode())
+                                        ? IndicesShardsStoresResponse.ShardStoreStatus.Allocation.PRIMARY
+                                        : IndicesShardsStoresResponse.ShardStoreStatus.Allocation.REPLICA;
+                            } else {
+                                allocation = IndicesShardsStoresResponse.ShardStoreStatus.Allocation.UNUSED;
+                            }
+                            shardStoreStatusList.add(new IndicesShardsStoresResponse.ShardStoreStatus(res.getNode(), res.version(), allocation, res.storeException()));
+                        }
+                        Collections.sort(shardStoreStatusList);
+                        shardStoreResponses.put(shardResponse.key, shardStoreStatusList);
+                    }
+                    indicesShardStoreStatuses.put(index, shardStoreResponses.build());
+                }
+                listener.onResponse(new IndicesShardsStoresResponse(indicesShardStoreStatuses.build(), failures.build()));
+            }
+
+            private boolean isPrimary(String index, int shardID, DiscoveryNode node) {
+                IndexRoutingTable indexRoutingTable = routingTables.index(index);
```

Comment:
I would use routing nodes here - it is keyed by node and will make the lookups more efficient (there is only one copy of a shard per node)